File: go/mechanisms/evm/exact/facilitator/eip3009.go
package facilitator

import (
	"context"
	"errors"
	"fmt"
	"math/big"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common"

	x402 "github.com/x402-foundation/x402/go/v2"
	"github.com/x402-foundation/x402/go/v2/mechanisms/evm"
	"github.com/x402-foundation/x402/go/v2/types"
)

// verifyEIP3009 verifies an EIP-3009 payment payload.
func (f *ExactEvmScheme) verifyEIP3009(
	ctx context.Context,
	payload types.PaymentPayload,
	requirements types.PaymentRequirements,
	simulate bool,
) (*x402.VerifyResponse, error) {
	if payload.Accepted.Scheme != evm.SchemeExact {
		return nil, x402.NewVerifyError(ErrInvalidScheme, "", fmt.Sprintf("invalid scheme: %s", payload.Accepted.Scheme))
	}

	if payload.Accepted.Network != requirements.Network {
		return nil, x402.NewVerifyError(ErrNetworkMismatch, "", fmt.Sprintf("network mismatch: %s != %s", payload.Accepted.Network, requirements.Network))
	}

	evmPayload, err := evm.PayloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewVerifyError(ErrInvalidPayload, "", fmt.Sprintf("failed to parse EVM payload: %s", err.Error()))
	}

	if evmPayload.Signature == "" {
		return nil, x402.NewVerifyError(ErrMissingSignature, "", "missing signature")
	}

	chainID, err := evm.GetEvmChainId(string(requirements.Network))
	if err != nil {
		return nil, x402.NewVerifyError(ErrFailedToGetNetworkConfig, "", err.Error())
	}

	tokenAddress := evm.NormalizeAddress(requirements.Asset)

	if !strings.EqualFold(evmPayload.Authorization.To, requirements.PayTo) {
		return nil, x402.NewVerifyError(ErrRecipientMismatch, "", fmt.Sprintf("recipient mismatch: %s != %s", evmPayload.Authorization.To, requirements.PayTo))
	}

	parsedAuthorization, err := ParseEIP3009Authorization(evmPayload.Authorization)
	if err != nil {
		return nil, x402.NewVerifyError(ErrInvalidPayload, evmPayload.Authorization.From, err.Error())
	}

	requiredValue, ok := new(big.Int).SetString(requirements.Amount, 10)
	if !ok {
		return nil, x402.NewVerifyError(ErrInvalidRequiredAmount, "", fmt.Sprintf("invalid required amount: %s", requirements.Amount))
	}

	if parsedAuthorization.Value.Cmp(requiredValue) != 0 {
		return nil, x402.NewVerifyError(ErrAuthorizationValueMismatch, evmPayload.Authorization.From, fmt.Sprintf("authorization value mismatch: %s != %s", parsedAuthorization.Value.String(), requiredValue.String()))
	}

	now := time.Now().Unix()
	if parsedAuthorization.ValidBefore.Cmp(big.NewInt(now+6)) < 0 {
		return nil, x402.NewVerifyError(ErrValidBeforeExpired, evmPayload.Authorization.From, fmt.Sprintf("valid before expired: %s", parsedAuthorization.ValidBefore.String()))
	}

	if parsedAuthorization.ValidAfter.Cmp(big.NewInt(now)) > 0 {
		return nil, x402.NewVerifyError(ErrValidAfterInFuture, evmPayload.Authorization.From, fmt.Sprintf("valid after in future: %s", parsedAuthorization.ValidAfter.String()))
	}

	tokenName, _ := requirements.Extra["name"].(string)
	tokenVersion, _ := requirements.Extra["version"].(string)
	if tokenName == "" || tokenVersion == "" {
		return nil, x402.NewVerifyError(ErrMissingEip712Domain, evmPayload.Authorization.From, "missing EIP-712 domain name/version in requirements.extra")
	}

	signatureBytes, err := evm.HexToBytes(evmPayload.Signature)
	if err != nil {
		return nil, x402.NewVerifyError(ErrInvalidSignatureFormat, evmPayload.Authorization.From, err.Error())
	}

	classification, err := ClassifyEIP3009Signature(
		ctx,
		f.signer,
		evmPayload.Authorization,
		signatureBytes,
		chainID,
		tokenAddress,
		tokenName,
		tokenVersion,
	)
	if err != nil {
		return nil, x402.NewVerifyError(ErrFailedToVerifySignature, evmPayload.Authorization.From, err.Error())
	}

	if !classification.Valid && classification.IsUndeployed && !HasEIP6492Deployment(classification.SigData) {
		return nil, x402.NewVerifyError(ErrUndeployedSmartWallet, evmPayload.Authorization.From, "")
	}

	if !classification.Valid && !classification.IsSmartWallet {
		return nil, x402.NewVerifyError(ErrInvalidSignature, evmPayload.Authorization.From, fmt.Sprintf("invalid signature: %s", evmPayload.Signature))
	}

	// Counterfactual ERC-6492 wallet: settle deploys via the factory, gated by the
	// allowlist. Enforce the same gate here so verify mirrors settle (a payment that
	// settle rejects with ErrFactoryNotAllowed must not verify as valid).
	if !classification.Valid && classification.IsUndeployed && HasEIP6492Deployment(classification.SigData) {
		if !evm.IsFactoryAllowed(classification.SigData.Factory, f.config.EIP6492AllowedFactories) {
			return nil, x402.NewVerifyError(ErrFactoryNotAllowed, evmPayload.Authorization.From, "factory not in EIP6492AllowedFactories allowlist")
		}
	}

	if errReason, err := evm.ValidateAssetIsContract(ctx, f.signer, requirements.Asset); err != nil {
		return nil, fmt.Errorf("asset contract check failed: %w", err)
	} else if errReason != "" {
		return nil, x402.NewVerifyError(errReason, evmPayload.Authorization.From, fmt.Sprintf("asset %s is not a deployed contract", requirements.Asset))
	}

	if simulate {
		simulationSucceeded, err := SimulateEIP3009Transfer(
			ctx,
			f.signer,
			tokenAddress,
			parsedAuthorization,
			classification.SigData,
		)
		if err != nil {
			return nil, x402.NewVerifyError(ErrEip3009SimulationFailed, evmPayload.Authorization.From, err.Error())
		}
		if !simulationSucceeded {
			reason := DiagnoseEIP3009SimulationFailure(
				ctx,
				f.signer,
				tokenAddress,
				evmPayload.Authorization,
				requiredValue,
				tokenName,
				tokenVersion,
			)
			return nil, x402.NewVerifyError(reason, evmPayload.Authorization.From, "")
		}
	}

	return &x402.VerifyResponse{
		IsValid: true,
		Payer:   evmPayload.Authorization.From,
	}, nil
}

// settleEIP3009 settles an EIP-3009 payment on-chain.
func (f *ExactEvmScheme) settleEIP3009(
	ctx context.Context,
	payload types.PaymentPayload,
	requirements types.PaymentRequirements,
	fctx *x402.FacilitatorContext,
) (*x402.SettleResponse, error) {
	network := x402.Network(payload.Accepted.Network)

	// Fast path: a prior settle attempt for this exact payload already broadcast
	// a transaction whose receipt wait failed (settlement_pending). The resource
	// server's single automatic retry resends the identical payload, so check the
	// pending-settlement store before re-verifying/re-broadcasting — reconcile
	// against the already-broadcast transaction instead of creating a second one.
	if evmPayload, parseErr := evm.PayloadFromMap(payload.Payload); parseErr == nil && evmPayload.Signature != "" {
		if txHash, ok, _ := f.pendingStore.Get(ctx, evmPayload.Signature); ok {
			// Remove before reconciling (rather than after) so a concurrent retry
			// of the same payload misses here instead of also reconciling: it
			// falls through to the normal broadcast path, which independently
			// rejects it as an on-chain replay (nonce already consumed).
			_ = f.pendingStore.Delete(ctx, evmPayload.Signature)
			return f.reconcilePendingEIP3009(ctx, evmPayload, requirements, network, txHash)
		}
	}

	verifyResp, err := f.verifyEIP3009(ctx, payload, requirements, f.config.SimulateInSettle)
	if err != nil {
		ve := &x402.VerifyError{}
		if errors.As(err, &ve) {
			return nil, x402.NewSettleError(ve.InvalidReason, ve.Payer, network, "", ve.InvalidMessage)
		}
		return nil, x402.NewSettleError(ErrVerificationFailed, "", network, "", err.Error())
	}

	evmPayload, err := evm.PayloadFromMap(payload.Payload)
	if err != nil {
		return nil, x402.NewSettleError(ErrInvalidPayload, verifyResp.Payer, network, "", err.Error())
	}

	tokenAddress := evm.NormalizeAddress(requirements.Asset)

	signatureBytes, err := evm.HexToBytes(evmPayload.Signature)
	if err != nil {
		return nil, x402.NewSettleError(ErrInvalidSignatureFormat, verifyResp.Payer, network, "", err.Error())
	}

	sigData, err := evm.ParseERC6492Signature(signatureBytes)
	if err != nil {
		return nil, x402.NewSettleError(ErrFailedToParseSignature, verifyResp.Payer, network, "", err.Error())
	}

	if HasEIP6492Deployment(sigData) {
		code, err := f.signer.GetCode(ctx, evmPayload.Authorization.From)
		if err != nil {
			return nil, x402.NewSettleError(ErrFailedToCheckDeployment, verifyResp.Payer, network, "", err.Error())
		}

		if len(code) == 0 {
			if !evm.IsFactoryAllowed(sigData.Factory, f.config.EIP6492AllowedFactories) {
				return nil, x402.NewSettleError(ErrFactoryNotAllowed, verifyResp.Payer, network, "", "")
			}

			if err := SendDeployTransaction(ctx, f.signer, sigData); err != nil {
				return nil, x402.NewSettleError(ErrSmartWalletDeploymentFailed, verifyResp.Payer, network, "", err.Error())
			}

			// Do NOT re-simulate the transfer here. The single authoritative pre-check is the
			// atomic deploy+transfer simulation that runs in verify (one eth_call via
			// Multicall3, state carried across both sub-calls). A second standalone eth_call
			// after the real deploy tx is unreliable — the read can race the deploy's state
			// propagation across load-balanced RPC nodes — and was producing false
			// inner-signature-unsupported rejections for valid wallets (e.g.
			// Coinbase Smart Wallet). The on-chain transferWithAuthorization below is the
			// definitive signature check; a genuinely unsupported inner signature reverts
			// there and is classified by parseEIP3009TransferError.
		}
	}

	parsedAuthorization, err := ParseEIP3009Authorization(evmPayload.Authorization)
	if err != nil {
		return nil, x402.NewSettleError(ErrInvalidPayload, verifyResp.Payer, network, "", err.Error())
	}

	dataSuffix, err := evm.ResolveDataSuffix(fctx, evm.DataSuffixContext{Payload: payload, Requirements: requirements})
	if err != nil {
		return nil, x402.NewSettleError(ErrInvalidPayload, verifyResp.Payer, network, "", err.Error())
	}

	txHash, err := ExecuteTransferWithAuthorization(ctx, f.signer, tokenAddress, parsedAuthorization, sigData, dataSuffix)
	if err != nil {
		return nil, x402.NewSettleError(parseEIP3009TransferError(err), verifyResp.Payer, network, "", err.Error())
	}

	return f.awaitEIP3009Settlement(ctx, evmPayload.Signature, tokenAddress, parsedAuthorization, network, verifyResp.Payer, txHash)
}

// reconcilePendingEIP3009 handles a pending-settlement store hit: it skips
// verify and broadcast entirely (the payer is taken directly from the
// payload, exactly as the original attempt did) and awaits the previously
// broadcast transaction.
func (f *ExactEvmScheme) reconcilePendingEIP3009(
	ctx context.Context,
	evmPayload *evm.ExactEIP3009Payload,
	requirements types.PaymentRequirements,
	network x402.Network,
	txHash string,
) (*x402.SettleResponse, error) {
	tokenAddress := evm.NormalizeAddress(requirements.Asset)
	parsedAuthorization, err := ParseEIP3009Authorization(evmPayload.Authorization)
	if err != nil {
		return nil, x402.NewSettleError(ErrInvalidPayload, evmPayload.Authorization.From, network, "", err.Error())
	}
	return f.awaitEIP3009Settlement(ctx, evmPayload.Signature, tokenAddress, parsedAuthorization, network, evmPayload.Authorization.From, txHash)
}

// awaitEIP3009Settlement waits for the broadcast transaction's receipt (via
// WaitForSettleReceiptWithPendingStore) and additionally verifies its
// Transfer event, shared by both the normal broadcast path and the
// pending-settlement reconciliation path above. A confirmed-but-mismatched
// receipt is terminal and clears the pending entry (unlike a receipt-wait
// failure, which WaitForSettleReceiptWithPendingStore already records for
// reconciliation); an unparseable-but-successful receipt re-records it as
// non-terminal, since the transfer's effect is unknown.
func (f *ExactEvmScheme) awaitEIP3009Settlement(
	ctx context.Context,
	pendingKey string,
	tokenAddress string,
	parsedAuthorization *ParsedEIP3009Authorization,
	network x402.Network,
	payer string,
	txHash string,
) (*x402.SettleResponse, error) {
	receipt, err := evm.WaitForSettleReceiptWithPendingStore(ctx, f.pendingStore, pendingKey, f.signer, txHash, payer, network,
		ErrTransactionFailed, ErrTransactionFailed)
	if err != nil {
		return nil, err
	}

	if receipt.Logs != nil {
		transferMatched, err := verifyEIP3009TransferEvent(receipt.Logs, common.HexToAddress(tokenAddress), expectedTransferEvent{
			From:  parsedAuthorization.From,
			To:    parsedAuthorization.To,
			Value: parsedAuthorization.Value,
		})
		if err != nil {
			// The receipt succeeded but its logs could not be parsed, so the transfer's effect
			// is unknown. A parsed-but-absent event below is terminal; this is not.
			if setErr := f.pendingStore.Set(ctx, pendingKey, txHash); setErr != nil {
				// Can't guarantee a later retry will find this to reconcile against — a
				// blind retry could re-verify/re-broadcast and double-send. Downgrade to
				// terminal, preserving the transaction hash for manual reconciliation.
				return nil, x402.NewSettleError(ErrTransactionFailed, payer, network, txHash,
					fmt.Sprintf("settlement_pending, but failed to persist for retry: %s", setErr.Error()))
			}
			return nil, x402.NewSettleError(ErrSettlementPending, payer, network, txHash,
				evm.TruncateErrorMessage(err.Error()))
		}
		if !transferMatched {
			_ = f.pendingStore.Delete(ctx, pendingKey)
			return nil, x402.NewSettleError(ErrTransferEventMismatch, payer, network, txHash, "")
		}
	}

	return &x402.SettleResponse{
		Success:     true,
		Transaction: txHash,
		Network:     network,
		Payer:       payer,
	}, nil
}

File: typescript/packages/mechanisms/evm/src/exact/facilitator/eip3009.ts
import {
  PaymentPayload,
  PaymentRequirements,
  FacilitatorContext,
  SettleResponse,
  VerifyResponse,
} from "@x402/core/types";
import { InMemoryPendingSettlementStore, PendingSettlementStore } from "@x402/core/facilitator";
import { getAddress, Hex, isAddressEqual, parseErc6492Signature } from "viem";
import { authorizationTypes } from "../../constants";
import { FacilitatorEvmSigner } from "../../signer";
import { getEvmChainId } from "../../utils";
import { ExactEIP3009Payload } from "../../types";
import * as Errors from "./errors";
import { resolveDataSuffix } from "../../shared/extensions";
import { verifyTypedDataSignature, classifyErc6492Payer } from "../../shared/verifySignature";
import {
  diagnoseEip3009SimulationFailure,
  executeTransferWithAuthorization,
  parseEip3009TransferError,
  simulateEip3009TransferResult,
  verifyEip3009TransferEvent,
} from "./eip3009-utils";
import {
  waitAndReturnSettleResponse,
  withPendingSettlementStore,
} from "../../shared/settleReceipt";

export interface VerifyEIP3009Options {
  /** Run onchain simulation. Defaults to true. */
  simulate?: boolean;
}

export interface EIP3009FacilitatorConfig {
  /**
   * Allowlist of factory contract addresses (hex strings, case-insensitive) that the facilitator
   * will call when deploying an undeployed smart wallet via ERC-6492.
   *
   * A non-empty list enables ERC-4337 smart wallet deployment via EIP-6492. Facilitators must
   * explicitly list every factory they trust to prevent arbitrary transaction injection via
   * attacker-controlled ERC-6492 signature wrappers. An empty or omitted list denies all factory
   * deployment calls.
   *
   * @default []
   */
  eip6492AllowedFactories?: string[];
  /**
   * If enabled, simulates transaction before settling. Defaults to false, ie only simulate during verify.
   *
   * @default false
   */
  simulateInSettle?: boolean;
}

/**
 * Verifies an EIP-3009 payment payload.
 *
 * @param signer - The facilitator signer for contract reads
 * @param payload - The payment payload to verify
 * @param requirements - The payment requirements
 * @param eip3009Payload - The EIP-3009 specific payload
 * @param options - Optional verification options
 * @param allowedFactories - Allowlisted ERC-6492 factory addresses; a counterfactual payment whose
 *   factory is not in this list is rejected here so verify mirrors settle's policy gate.
 * @returns Promise resolving to verification response
 */
export async function verifyEIP3009(
  signer: FacilitatorEvmSigner,
  payload: PaymentPayload,
  requirements: PaymentRequirements,
  eip3009Payload: ExactEIP3009Payload,
  options?: VerifyEIP3009Options,
  allowedFactories: string[] = [],
): Promise<VerifyResponse> {
  const payer = eip3009Payload.authorization.from;
  let eip6492Deployment:
    | { factoryAddress: `0x${string}`; factoryCalldata: `0x${string}` }
    | undefined;

  // Verify scheme matches
  if (payload.accepted.scheme !== "exact" || requirements.scheme !== "exact") {
    return {
      isValid: false,
      invalidReason: Errors.ErrInvalidScheme,
      payer,
    };
  }

  // Get chain configuration
  if (!requirements.extra?.name || !requirements.extra?.version) {
    return {
      isValid: false,
      invalidReason: Errors.ErrMissingEip712Domain,
      payer,
    };
  }

  const { name, version } = requirements.extra as { name: string; version: string };
  const erc20Address = getAddress(requirements.asset);

  // Verify network matches
  if (payload.accepted.network !== requirements.network) {
    return {
      isValid: false,
      invalidReason: Errors.ErrNetworkMismatch,
      payer,
    };
  }

  // Build typed data for signature verification
  const permitTypedData = {
    types: authorizationTypes,
    primaryType: "TransferWithAuthorization" as const,
    domain: {
      name,
      version,
      chainId: getEvmChainId(requirements.network),
      verifyingContract: erc20Address,
    },
    message: {
      from: eip3009Payload.authorization.from,
      to: eip3009Payload.authorization.to,
      value: BigInt(eip3009Payload.authorization.value),
      validAfter: BigInt(eip3009Payload.authorization.validAfter),
      validBefore: BigInt(eip3009Payload.authorization.validBefore),
      nonce: eip3009Payload.authorization.nonce,
    },
  };

  const signature = eip3009Payload.signature!;

  // Classify the payer: fetch code once, parse ERC-6492 wrapper, determine counterfactual.
  // Using classifyErc6492Payer avoids duplicating this block across eip3009.ts / v1/scheme.ts.
  const {
    isCounterfactual,
    innerSignature,
    eip6492Deployment: classification6492,
  } = await classifyErc6492Payer(signer, signature, payer);

  if (classification6492) {
    eip6492Deployment = classification6492;
  }

  if (isCounterfactual) {
    // Counterfactual deposits are deployed by settle via the factory, which is gated by the
    // allowlist. Enforce the same gate here so verify does not return isValid:true for a payment
    // that settle will reject with ErrFactoryNotAllowed (verify must predict settle).
    const factory = classification6492?.factoryAddress;
    const factoryAllowed =
      !!factory && allowedFactories.some(a => a.trim().toLowerCase() === factory.toLowerCase());
    if (!factoryAllowed) {
      return {
        isValid: false,
        invalidReason: Errors.ErrFactoryNotAllowed,
        payer,
      };
    }
  }

  if (!isCounterfactual) {
    // For deployed addresses (plain EOA, smart contract, 7702 EOA): verify the
    // signature using a strict primitive that mirrors on-chain SignatureChecker
    // semantics — ecrecover when no code, strict EIP-1271 when code is present.
    // No ECDSA fallback for code addresses; that fallback causes pre-verify to
    // accept sigs the on-chain token rejects (empirically confirmed on Base Sepolia).
    const isValid = await verifyTypedDataSignature(signer, {
      address: eip3009Payload.authorization.from,
      ...permitTypedData,
      signature: innerSignature,
    });
    if (!isValid) {
      return {
        isValid: false,
        invalidReason: Errors.ErrInvalidSignature,
        payer,
      };
    }
  }
  // Counterfactual: skip pre-verify and defer to on-chain simulation/settle which
  // deploys the factory first then atomically validates the signature.

  // Verify payment recipient matches
  if (getAddress(eip3009Payload.authorization.to) !== getAddress(requirements.payTo)) {
    return {
      isValid: false,
      invalidReason: Errors.ErrRecipientMismatch,
      payer,
    };
  }

  // Verify validBefore is in the future (with 6 second buffer for block time)
  const now = Math.floor(Date.now() / 1000);
  if (BigInt(eip3009Payload.authorization.validBefore) < BigInt(now + 6)) {
    return {
      isValid: false,
      invalidReason: Errors.ErrValidBeforeExpired,
      payer,
    };
  }

  // Verify validAfter is not in the future
  if (BigInt(eip3009Payload.authorization.validAfter) > BigInt(now)) {
    return {
      isValid: false,
      invalidReason: Errors.ErrValidAfterInFuture,
      payer,
    };
  }

  // Verify amount exactly matches requirements
  if (BigInt(eip3009Payload.authorization.value) !== BigInt(requirements.amount)) {
    return {
      isValid: false,
      invalidReason: Errors.ErrAuthorizationValueMismatch,
      payer,
    };
  }

  // Reject payments whose asset is an EOA — eth_call on an EOA silently returns
  // empty data without reverting, so simulation would pass but no Transfer event
  // would be emitted, producing a silent no-op settlement.
  const assetBytecode = await signer.getCode({ address: erc20Address });
  if (!assetBytecode || assetBytecode === "0x") {
    return { isValid: false, invalidReason: Errors.ErrAssetNotDeployedContract, payer };
  }

  // Transaction simulation
  if (options?.simulate !== false) {
    const { ok, error: simError } = await simulateEip3009TransferResult(
      signer,
      erc20Address,
      eip3009Payload,
      eip6492Deployment,
    );
    if (!ok) {
      const diagnosis = await diagnoseEip3009SimulationFailure(
        signer,
        erc20Address,
        eip3009Payload,
        requirements,
        requirements.amount,
      );
      // Carry the raw revert text so the concrete reason survives the mapping to a code.
      const rawMessage =
        simError instanceof Error ? simError.message : simError ? String(simError) : undefined;
      return rawMessage ? { ...diagnosis, invalidMessage: rawMessage } : diagnosis;
    }
  }

  return {
    isValid: true,
    invalidReason: undefined,
    payer,
  };
}

/**
 * Waits for the broadcast transaction's receipt (and, when logs are present,
 * verifies its Transfer event), shared by both the normal broadcast path and
 * the pending-settlement reconciliation path in {@link settleEIP3009}.
 * Delegates `PendingSettlementStore` bookkeeping to
 * {@link withPendingSettlementStore}: only a `settlement_pending` outcome
 * (a receipt-wait failure, or a confirmed receipt whose logs couldn't be
 * parsed) is recorded; a confirmed receipt with a mismatched Transfer event
 * is terminal and clears the entry instead.
 *
 * @param signer - The facilitator signer for the receipt wait
 * @param store - Pending-settlement store keyed by the EIP-3009 signature
 * @param pendingKey - The EIP-3009 signature used as the store key
 * @param tx - The broadcast transaction hash to await
 * @param network - The network the transaction was broadcast to
 * @param payer - The payer address
 * @param asset - The ERC-20 asset address (checksummed)
 * @param auth - The authorization fields used to validate the Transfer event
 * @param auth.from - Authorization sender
 * @param auth.to - Authorization recipient
 * @param auth.value - Authorization value (atomic units)
 * @returns Promise resolving to the settlement response
 */
async function awaitEIP3009Settlement(
  signer: FacilitatorEvmSigner,
  store: PendingSettlementStore,
  pendingKey: string,
  tx: `0x${string}`,
  network: PaymentRequirements["network"],
  payer: string,
  asset: `0x${string}`,
  auth: { from: string; to: string; value: string },
): Promise<SettleResponse> {
  return withPendingSettlementStore(
    store,
    pendingKey,
    () =>
      waitAndReturnSettleResponse(signer, tx, network, payer, {
        failedStatusReason: Errors.ErrTransactionFailed,
        validateReceipt: receipt => {
          if (
            receipt.logs != null &&
            !verifyEip3009TransferEvent(receipt.logs, asset, {
              from: getAddress(auth.from),
              to: getAddress(auth.to),
              value: BigInt(auth.value),
            })
          ) {
            return {
              success: false,
              errorReason: Errors.ErrTransferEventMismatch,
              transaction: tx,
              network,
              payer,
            };
          }
          return undefined;
        },
      }),
    Errors.ErrTransactionFailed,
  );
}

/**
 * Settles an EIP-3009 payment by executing transferWithAuthorization.
 *
 * @param signer - The facilitator signer for contract writes
 * @param payload - The payment payload to settle
 * @param requirements - The payment requirements
 * @param eip3009Payload - The EIP-3009 specific payload
 * @param config - Facilitator configuration
 * @param context - Optional facilitator context for extension capabilities
 * @param store - Pending-settlement store. A prior settle attempt for this exact
 *   payload that broadcast a transaction whose receipt wait failed
 *   (settlement_pending) is reconciled against here instead of re-verifying and
 *   re-broadcasting. Defaults to a fresh in-memory store when omitted (no
 *   cross-call sharing).
 * @returns Promise resolving to settlement response
 */
export async function settleEIP3009(
  signer: FacilitatorEvmSigner,
  payload: PaymentPayload,
  requirements: PaymentRequirements,
  eip3009Payload: ExactEIP3009Payload,
  config: EIP3009FacilitatorConfig,
  context?: FacilitatorContext,
  store: PendingSettlementStore = new InMemoryPendingSettlementStore(),
): Promise<SettleResponse> {
  const payer = eip3009Payload.authorization.from;
  const signature = eip3009Payload.signature;

  // Fast path: a prior settle attempt for this exact payload already broadcast
  // a transaction whose receipt wait failed (settlement_pending). The resource
  // server's single automatic retry resends the identical payload, so check the
  // pending-settlement store before re-verifying/re-broadcasting — reconcile
  // against the already-broadcast transaction instead of creating a second one.
  if (signature) {
    const cachedTx = await store.get(signature);
    if (cachedTx) {
      // Remove before reconciling (rather than after) so a concurrent retry
      // of the same payload misses here instead of also reconciling: it
      // falls through to the normal broadcast path, which independently
      // rejects it as an on-chain replay (nonce already consumed).
      await store.delete(signature);
      return awaitEIP3009Settlement(
        signer,
        store,
        signature,
        cachedTx as `0x${string}`,
        payload.accepted.network,
        payer,
        getAddress(requirements.asset),
        eip3009Payload.authorization,
      );
    }
  }

  // Re-verify before settling
  const valid = await verifyEIP3009(
    signer,
    payload,
    requirements,
    eip3009Payload,
    { simulate: config.simulateInSettle ?? false },
    config.eip6492AllowedFactories ?? [],
  );
  if (!valid.isValid) {
    return {
      success: false,
      network: payload.accepted.network,
      transaction: "",
      errorReason: valid.invalidReason ?? Errors.ErrInvalidScheme,
      payer,
    };
  }

  try {
    // Parse ERC-6492 signature if applicable (for optional deployment).
    // Keep the full result so we can access the inner signature later for
    // the post-deploy transfer simulation.
    const settleErc6492Data = parseErc6492Signature(signature!);
    const {
      address: factoryAddress,
      data: factoryCalldata,
      signature: erc6492InnerSig,
    } = settleErc6492Data;

    // Deploy ERC-4337 smart wallet via EIP-6492 if factory is in the allowlist
    if (
      factoryAddress &&
      factoryCalldata &&
      !isAddressEqual(factoryAddress, "0x0000000000000000000000000000000000000000")
    ) {
      // Check if smart wallet is already deployed
      const bytecode = await signer.getCode({ address: payer });

      if (!bytecode || bytecode === "0x") {
        const normalizedFactory = factoryAddress.toLowerCase();
        const isAllowed = (config.eip6492AllowedFactories ?? []).some(
          allowed => allowed.toLowerCase() === normalizedFactory,
        );
        if (!isAllowed) {
          return {
            success: false,
            errorReason: Errors.ErrFactoryNotAllowed,
            transaction: "",
            network: payload.accepted.network,
            payer,
          };
        }

        // Wallet not deployed - attempt deployment
        const deployTx = await signer.sendTransaction({
          to: factoryAddress as Hex,
          data: factoryCalldata as Hex,
        });

        // Wait for deployment and check whether it actually succeeded.
        // A reverted factory tx would silently proceed without this check, misclassifying
        // the downstream transfer failure as an invalid-signature error.
        const deployReceipt = await signer.waitForTransactionReceipt({ hash: deployTx });
        if (deployReceipt.status !== "success") {
          return {
            success: false,
            errorReason: Errors.ErrSmartWalletDeploymentFailed,
            transaction: "",
            network: payload.accepted.network,
            payer,
          };
        }

        // Do NOT re-simulate the transfer here. The single authoritative pre-check is the
        // atomic Multicall3 deploy+transfer simulation that runs in verify (one eth_call,
        // state carried across both sub-calls). A second standalone eth_call after the real
        // deploy tx is unreliable — the read can race the deploy's state propagation across
        // load-balanced RPC nodes — and was producing false `eip6492_deployed_inner_wallet_
        // signature_unsupported` rejections for valid wallets (e.g. Coinbase Smart Wallet).
        // The on-chain transferWithAuthorization below is itself the definitive signature
        // check (the token routes to the wallet's isValidSignature); a genuinely
        // unsupported inner signature reverts there and is classified by the catch block.
      }
    }

    const dataSuffix = await resolveDataSuffix(context, {
      paymentPayload: payload,
      paymentRequirements: requirements,
    });

    // When the original signature was ERC-6492 wrapped (deployed wallet), use the
    // extracted inner signature for the on-chain transferWithAuthorization call.
    // FiatTokenV2_2's isValidSignature on the deployed contract expects the compact inner signature
    const settlePayload =
      erc6492InnerSig && erc6492InnerSig !== signature
        ? { ...eip3009Payload, signature: erc6492InnerSig }
        : eip3009Payload;

    const asset = getAddress(requirements.asset);
    const tx = await executeTransferWithAuthorization(signer, asset, settlePayload, dataSuffix);

    // Receipt status only proves the tx did not revert.
    // When logs are present, require the expected ERC-20 Transfer event.
    if (signature) {
      return await awaitEIP3009Settlement(
        signer,
        store,
        signature,
        tx,
        payload.accepted.network,
        payer,
        asset,
        eip3009Payload.authorization,
      );
    }
    return await waitAndReturnSettleResponse(signer, tx, payload.accepted.network, payer, {
      failedStatusReason: Errors.ErrTransactionFailed,
      validateReceipt: receipt => {
        if (
          receipt.logs != null &&
          !verifyEip3009TransferEvent(receipt.logs, asset, {
            from: getAddress(eip3009Payload.authorization.from),
            to: getAddress(eip3009Payload.authorization.to),
            value: BigInt(eip3009Payload.authorization.value),
          })
        ) {
          return {
            success: false,
            errorReason: Errors.ErrTransferEventMismatch,
            transaction: tx,
            network: payload.accepted.network,
            payer,
          };
        }
        return undefined;
      },
    });
  } catch (error) {
    // Preserve the raw revert text alongside the mapped code. The mapper collapses many
    // distinct on-chain reverts into a single reason (e.g. ErrInvalidSignature), so without
    // the original message the true cause is invisible to callers/operators.
    return {
      success: false,
      errorReason: parseEip3009TransferError(error),
      errorMessage: error instanceof Error ? error.message : String(error),
      transaction: "",
      network: payload.accepted.network,
      payer,
    };
  }
}
